Support copyToRoot paths that are symlinks to other store paths

sysedwinistrator opened the pull request.

sysedwinistrator committed:
Support copyToRoot paths that are symlinks to other store paths

elpdt852 commented on pkg/nix2container/generate.go:
```diff
@@ -290,11 +293,31 @@ func writeNixClosureLayer(ctx context.Context, w io.Writer, nixStorePaths, copyT
+						return err
+					}
+
+					storePathRegex := regexp.MustCompile("/nix/store/[0-9a-z]{32}-[-.+_0-9a-zA-Z]+")
```
Due to $DAYJOB non-standard nix store dir, I'd like nix-snapshotter to not have `/nix/store` assumptions. Is there a way we can support symlinks without assuming store paths or pass in a storeDir as an option which is defaulted to `/nix/store`?

sysedwinistrator replied:
Probably.

So just for clarification, the regexp is there to check that the symlink target is itself another top-level store path (and it can only be a directory, but that's checked in the recursive call).
It's not about verifying that the symlink target's is inside the nix store. For regular (non-symlink) `copyToRoot` paths, that's currently not even checked at all (only whether the path is an directory).

I see two solutions to do this without having to manually specify the nix store location:
- Build the correct regex by asking the Nix daemon what the nix store location is (I don't know how to do that, at least `nix store info --json` doesn't print that information). Ideally, we should only do this once (and I now realize `storePathRegex` should probably be defined only once as well)
- Ask the Nix daemon whether the target is a top-level path. I don't think you can do this directly, but `nix path-info --json <store path>` always returns the top-level info even if `<store path>` is a subpath, so we can just check whether the path info we got back is for our exact store path.

sysedwinistrator replied:
Actually, Nix sets the `NIX_STORE` environment variable to the store dir for every derivation: https://github.com/NixOS/nix/blob/73d3ab05b65c2ffa24b22e206a48617b53bff98e/src/libstore/unix/build/derivation-builder.cc#L1006

Assuming nix2container is only used in derivations (and we could just default the variable to `/nix/store` for other use cases), that solution is much simpler than connecting the nix daemon (which seems like a bad idea to do from a derivation, if it's possible at all).

sysedwinistrator committed:
Only define store path regex once when creating closure

sysedwinistrator committed:
Support custom nix store location

Nix sets NIX_STORE to the nix store dir (e.g. /nix/store) in the
derivation builder:
https://github.com/NixOS/nix/blob/73d3ab05b65c2ffa24b22e206a48617b53bff98e/src/libstore/unix/build/derivation-builder.cc#L1006
